src/ast/visitor.rs

```diff
@@ -0,0 +1,148 @@
+use crate::{
+    ast::{BinOp, Block, Literal, Located, Name, Node, UnOp},
+    ty::{Binding, Ty},
+};
+
+pub trait NodeVisitor<'a> {
+    fn super_block(&mut self, block: &Block<'a>) {
+        for node in block {
+            self.visit_node(&node);
+        }
+    }
+
+    fn super_node(&mut self, node: &Located<Node<'a>>) {
+        match &node.content {
+            Node::BinaryOp(op, node1, node2) => {
+                self.visit_binary_op(&op, node1.as_ref(), node2.as_ref())
+            }
+            Node::UnaryOp(op, node) => self.visit_unary_op(&op, node.as_ref()),
+            Node::LetBind(name, opt_ty, node) => self.visit_let_bind(&name, &opt_ty, node.as_ref()),
+            Node::Cond(if_blk, do_blk, el_blk) => self.visit_cond(&if_blk, &do_blk, &el_blk),
+            Node::FnDef(opt_name, args, body, opt_ty) => {
+                self.visit_fn_def(&opt_name, &args, &body, &opt_ty)
+            }
+            Node::Call(func, args) => self.visit_call(&func, &args),
+            Node::Literal(literal) => self.visit_literal(&literal),
+            Node::Name(name) => self.visit_name(&name),
+        }
+    }
+
+    fn super_binary_op(
+        &mut self,
+        _op: &BinOp,
+        node1: &Located<Node<'a>>,
+        node2: &Located<Node<'a>>,
+    ) {
+        self.visit_node(node1);
+        self.visit_node(node2);
+    }
+
+    fn super_unary_op(&mut self, _op: &UnOp, node: &Located<Node<'a>>) {
+        self.visit_node(node);
+    }
+
+    fn super_let_bind(
+        &mut self,
+        name: &Located<Name<'a>>,
+        _opt_ty: &Option<Located<Ty>>,
+        node: &Located<Node<'a>>,
+    ) {
+        self.visit_name(&name.content);
+        self.visit_node(node);
+    }
+
+    fn super_cond(
+        &mut self,
+        if_blk: &Located<Block<'a>>,
+        do_blk: &Located<Block<'a>>,
+        el_blk: &Located<Block<'a>>,
+    ) {
+        self.visit_block(&if_blk.content);
+        self.visit_block(&do_blk.content);
+        self.visit_block(&el_blk.content);
+    }
+
+    fn super_fn_def(
+        &mut self,
+        opt_name: &Option<Located<Name<'a>>>,
+        _args: &Vec<Located<Binding<'a>>>,
```

Comment:
Instead of `&Vec` a slice reference is usually better `&[]`

Reply:
Is it there any other reason for that besides coercion? If someone needs coercion there probably they're misusing the trait.

Reply:
Oh there is an extra reference (of course, vecs have an slice inside them). I'll change this.